staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go

```diff
@@ -0,0 +1,150 @@
+	GetApiServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
+}
+
+// OpenAPIAggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove
+// them if necessary.
+type OpenAPIAggregationController struct {
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+	downloader                *openAPIDownloader
+}
+
+func NewOpenAPIAggregationController(downloader *openAPIDownloader, openAPIAggregationManager OpenAPIAggregationManager) *OpenAPIAggregationController {
+	c := &OpenAPIAggregationController{
+		openAPIAggregationManager: openAPIAggregationManager,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(successfulUpdateDelay, failedUpdateMaxExpDelay), "APIServiceOpenAPIAggregationControllerQueue1"),
+		downloader: downloader,
+	}
+
+	return c
+}
+
+func (c *OpenAPIAggregationController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
+	defer glog.Infof("Shutting down OpenAPIAggregationController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *OpenAPIAggregationController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	defer c.queue.Done(key)
+	if quit {
+		return false
+	}
+	apiServiceName := key.(string)
+	handler, etag, exists := c.openAPIAggregationManager.GetApiServiceInfo(apiServiceName)
+	if !exists || handler == nil {
+		c.queue.Forget(key)
+		return true
+	}
+	returnSpec, newEtag, httpStatus, err := c.downloader.Download(handler, etag)
+	rateLimit := false
+	switch {
+	case err != nil:
+		utilruntime.HandleError(fmt.Errorf("loading OpenAPI spec for %q failed with : %v", apiServiceName, err))
+		rateLimit = true
+	case httpStatus == http.StatusNotModified:
+		rateLimit = false
+	case httpStatus == http.StatusNotFound || returnSpec == nil:
+		rateLimit = true
+	case httpStatus == http.StatusOK:
+		if err := c.openAPIAggregationManager.UpdateApiServiceSpec(apiServiceName, returnSpec, newEtag); err != nil {
+			utilruntime.HandleError(fmt.Errorf("aggregating OpenAPI spec for %q failed with : %v", apiServiceName, err))
+			rateLimit = true
+		}
+	}
+
+	if rateLimit {
+		c.queue.AddRateLimited(key)
+	} else {
+		c.queue.Forget(key)
+		c.queue.AddAfter(key, successfulUpdateDelay)
+	}
+	return true
+}
+
+func (c *OpenAPIAggregationController) AddAPIService(handler http.Handler, apiService *apiregistration.APIService) {
+	if err := c.openAPIAggregationManager.AddUpdateApiService(handler, apiService); err != nil {
+		utilruntime.HandleError(fmt.Errorf("adding %q to OpenAPIAggregationController failed with : %v", apiService.Name, err))
```

Comment:
nit: drop space between between with and ':'. 

So, `with:`.

Ditto for the two other error messages in this file.

Reply:
Fixed.

Reply:
Done in next commit.